Interactivity API: Add JSDoc comments to the public API

DAreRodz opened the pull request:
## What?

Adds JSDoc comments to the `store` and `directive` functions, which are currently the public part of the Interactivity API.

## Why?

To start documenting the Interactivity API.

## How?

I followed the format used for other JSDoc comments in the repository. I added basic support for typing, but it is very preliminary and not meant to be fully compatible with TypeScript yet.

DAreRodz committed:
Add initial JSDoc for `directive`

DAreRodz committed:
Add initial JSDoc for `store`

DAreRodz committed:
Improve example for `store()`

DAreRodz committed:
Improve store docs a little

DAreRodz committed:
Rename Element type to VNode

luisherranz commented:
@darerodz, is this ready for review?

DAreRodz commented:
Yes, it is. 🙏

luisherranz commented on packages/interactivity/src/store.js:
```diff
@@ -38,6 +38,47 @@ const getSerializedState = () => {
+ * @param {Object} properties             Properties to be added to the global store.
+ * @param {Object} [properties.state]     State to be added to the global store.
+ * @param {Object} [properties.selectors] Selectors to be added to the global store.
+ * @param {Object} [properties.actions]   Actions to be added to the global store.
```
For now, this are arbitraty names. Only state is special. So maybe we shouldn't have special mentions for them.

```suggestion
```

luisherranz commented on packages/interactivity/src/store.js:
```diff
@@ -38,6 +38,47 @@ const getSerializedState = () => {
+ * Extends the Interactivity API global store with the passed properties.
+ *
+ * These props tipically consist of `state`, `selectors` and `actions` that can
+ * be used by any directive, making the HTML reactive and interactive.
```
Maybe we can separate `state` from the others and mention that it is reactive.

```suggestion
 * These props typically consist of `state`, which is reactive, and other
 * properties like `selectors`, `actions`, `effects`, etc. which can store
 * callbacks and derived state. These props can then be referenced by any
 * directive to make the HTML interactive.
```

luisherranz commented on packages/interactivity/src/hooks.js:
```diff
@@ -8,14 +8,71 @@ import { useRef, useMemo } from 'preact/hooks';
+ * directive(
+ *   'alert', // Name without the `data-wp-` prefix.
+ *   ( { directives: { alert }, element, evaluate }) => {
+ *     element.props.onClick = () => {
```
[We're using `onclick` instead of `onClick`](https://github.com/WordPress/block-interactivity-experiments/blob/main-wp-directives-plugin/src/runtime/directives.js#L162) in `wp-link`. I don't know if both are valid, probably, but it'd be good to check before showing it in the example.

```suggestion
 *     element.props.onclick = () => {
```

luisherranz commented on packages/interactivity/src/hooks.js:
```diff
@@ -8,14 +8,71 @@ import { useRef, useMemo } from 'preact/hooks';
+ *   'alert', // Name without the `data-wp-` prefix.
+ *   ( { directives: { alert }, element, evaluate }) => {
+ *     element.props.onClick = () => {
+ *       alert( evaluate( alert.default ) );
```
Would it be worth to also explain the suffix part? Like `data-wp-alert--some-suffix`.

DAreRodz replied:
I think it would. 🤔 I'll try and come up with an example.

DAreRodz replied:
Added in https://github.com/WordPress/gutenberg/pull/52469/commits/4ec8dbb361cbd66659a08dca84536a3b4b7ce6b5.

luisherranz reviewed (commented):
Thanks, David. I like the examples 🙂

I left a few comments below. Let me know what you think.

luisherranz commented:
@darerodz, is anything blocking the merge of this?

DAreRodz committed:
Remove selectors and actions from store types

Co-authored-by: Luis Herranz <luisherranz@gmail.com>

DAreRodz committed:
Clarify that `state` is reactive

Co-authored-by: Luis Herranz <luisherranz@gmail.com>

DAreRodz committed:
Replace `onClick` with `onclick` in the JSDoc example

Co-authored-by: Luis Herranz <luisherranz@gmail.com>

DAreRodz commented:
> is anything blocking the merge of this?

Hey, @luisherranz; sorry for the delay. I prioritized other things, but I'm now addressing your feedback. This PR will be ready to merge in a moment.

DAreRodz committed:
Fix `onclick` in JSDoc explanation

luisherranz commented:
Thanks. Let me know once it's ready for another review 🙂

DAreRodz committed:
Add example with custom suffixes

DAreRodz commented:
@luisherranz, it's ready for review now.

EDIT: I had to fix a jslint error that was accidentally included.

DAreRodz committed:
Fix jslint errors

DAreRodz commented on packages/interactivity/src/store.js:
```diff
@@ -38,6 +38,49 @@ const getSerializedState = () => {
+ * @param {Object} properties         Properties to be added to the global store.
+ * @param {Object} [properties.state] State to be added to the global store. All
+ *                                    the properties included here become reactive.
+ * @param {Object} [properties.block] Other properties to be added to the global store.
```
I believe this is not the correct way of documenting a variable that uses the spread operator. It seems to work somehow, but only for the docs, not for types. We could maybe remove it.

luisherranz replied:
Up to you 🙂

DAreRodz replied:
I'll remove it for now, then.

luisherranz commented on packages/interactivity/src/hooks.js:
````diff
@@ -8,12 +8,104 @@ import { useRef, useCallback } from 'preact/hooks';
+ * )
+ * ```
+ *
+ * The previous code register a custom directive type for displaying an alert
````
```suggestion
 * The previous code registers a custom directive type for displaying an alert
```

luisherranz commented on packages/interactivity/src/hooks.js:
````diff
@@ -8,12 +8,104 @@ import { useRef, useCallback } from 'preact/hooks';
+ * ```html
+ * <button data-wp-alert="state.messages.alert">Click me!</button>
+ * ```
+ * Note that, in the previous example, you acces `alert.default` in order to
````
```suggestion
 * Note that, in the previous example, you access `alert.default` in order to
```

luisherranz reviewed (approved):
LGTM, David!

luisherranz committed:
Fix a couple of typos

luisherranz commented:
Maybe we can add an `Enhancement` entry in the changelog for this.

DAreRodz committed:
Remove docs for dubious properties.block

DAreRodz committed:
Add changelog
